From 74295e131f112076be60744b6d38327ac8e0894c Mon Sep 17 00:00:00 2001
From: Sreenidhi B R <sreenira@broadcom.com>
Date: Tue, 30 Jul 2013 14:26:52 +0530
Subject: [PATCH 1170/1532] dtre: modify dtre module for xlp2xx

- only two channels (instead of usual four)
- only DMA support (RAID not supported)
[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
---
 drivers/misc/netlogic/dtre/nlm_adma.c | 24 ++++++++++++++++++++----
 drivers/misc/netlogic/dtre/nlm_adma.h |  1 +
 2 files changed, 21 insertions(+), 4 deletions(-)

diff --git a/drivers/misc/netlogic/dtre/nlm_adma.c b/drivers/misc/netlogic/dtre/nlm_adma.c
index 8ef95d9..8ca2167 100644
--- a/drivers/misc/netlogic/dtre/nlm_adma.c
+++ b/drivers/misc/netlogic/dtre/nlm_adma.c
@@ -434,6 +434,10 @@ static dma_cookie_t nlm_tx_submit (struct dma_async_tx_descriptor *tx)
 	      channel 0,1: belong to DTRE-0 and channel 0 is for RAID
 	      channel 2,3: belong to DTRE-1 and channel 2 is for RAID
 
+	      for xlp2xx:
+	      there is no RAID support
+	      channel0-1, for DMA.
+
 	   others:
 	      channel 0: RAID
 	      channel 1-3: DMA
@@ -1465,7 +1469,8 @@ static int __devinit nlm_adma_probe(struct platform_device *pdev)
 	int i, err, loop;
 	uint64_t base;
 	uint32_t value;
-	int is_raid = 0;
+	int num_channels;
+	volatile int is_raid = 0;
 
 	if (is_nlm_xlp9xx())
 	{
@@ -1494,7 +1499,14 @@ static int __devinit nlm_adma_probe(struct platform_device *pdev)
 		}
 	}
 
-	for (loop = 0; loop < DTRE_NUM_CHANNELS; loop++)
+	if (is_nlm_xlp2xx()) {
+		num_channels = DTRE_2XX_NUM_CHANNELS;
+	}
+	else {
+		num_channels = DTRE_NUM_CHANNELS;
+	}
+
+	for (loop = 0; loop < num_channels; loop++)
 	{
 		/* assume current channel is DMA (not raid)
 		   change it later in the loop (if required) */
@@ -1553,6 +1565,10 @@ static int __devinit nlm_adma_probe(struct platform_device *pdev)
 				is_raid = 0;
 			}
 		}
+		else if (is_nlm_xlp2xx()) {
+			/* for xlp2xx, there are TWO channels, both for DMA */
+			is_raid = 0;
+		}
 		else
 		{
 			/* set RAID capabilities for chan 0 and 
@@ -1598,7 +1614,7 @@ static int __devinit nlm_adma_probe(struct platform_device *pdev)
 		dma_dev->dev = &pdev->dev;
 		dma_async_device_register(dma_dev);
 
-	} /* end of loop DTRE_NUM_CHANNELS */
+	} /* end of loop num_channels */
 
 	nlm_dtre_null_page = alloc_page(GFP_KERNEL);
 	if (nlm_dtre_null_page == NULL) {
@@ -1634,7 +1650,7 @@ static int __devinit nlm_adma_probe(struct platform_device *pdev)
 		/* set channel 0 is raid - bit 4 */
 		nlm_hal_write_32bit_reg (base, 0x40, value | (0x10));
 	}
-	else
+	else if (!is_nlm_xlp2xx())
 	{
 		/* for DTRE, B/D/F = 0/5/0 */
 		base = nlm_hal_get_dev_base (0 /*node*/, 0 /*B*/, 5 /*D*/, 0 /*F*/);
diff --git a/drivers/misc/netlogic/dtre/nlm_adma.h b/drivers/misc/netlogic/dtre/nlm_adma.h
index e4e0b13..77e73e2 100644
--- a/drivers/misc/netlogic/dtre/nlm_adma.h
+++ b/drivers/misc/netlogic/dtre/nlm_adma.h
@@ -55,6 +55,7 @@
 #define DTRE_MIN_VC		264
 #define DTRE_MAX_VC		267
 #define DTRE_NUM_CHANNELS	4
+#define DTRE_2XX_NUM_CHANNELS	2
 #define DTRE_9XX_MIN_VC		392
 #define DTRE_9XX_MAX_VC		395
 
-- 
1.9.1

